src/main/scala/chisel3/util/experimental/ForceNames.scala

```diff
@@ -0,0 +1,290 @@
+    path.map { case (inst, of) => s"${inst.value}:${of.value}" }.mkString("/")
+}
+
+/** Forces the name of marked signals to a certain name
+  *   - If there is a conflict in the enclosing module's namespace, throws an exception
+  *   - Renames signals of ground types only. If you rename an intermediate module, it will throw an error
+  *   - Renames instances as well (as well as port names)
+  * Common usages:
+  *   - Use to avoid prefixing behavior on specific instances whose enclosing modules are inlined
+  */
+class ForceNamesTransform extends Transform with DependencyAPIMigration {
+  override def optionalPrerequisites: Seq[TransformDependency] = Seq(Dependency[InlineInstances])
+  override def optionalPrerequisiteOf: Seq[TransformDependency] = Forms.LowEmitters
+  override def prerequisites: Seq[TransformDependency] = Seq(Dependency(LowerTypes))
+  override def invalidates(a: Transform): Boolean = firrtl.passes.InferTypes == a
+
+  import ForceNamesTransform._
+
+  /** Renames signals/instances in a module
+    *
+    * @throws FirrtlUserException
+    * @param modToNames Maps module name to map of old signal/instance name to new signal/instance name
+    * @param renameMap Record renames
+    * @param ct Circuit target
+    * @param igraph InstanceKeyGraph of parent circuit
+    * @param mod Enclosing module of the names to force
+    * @return
+    */
+  private def forceNamesInModule(
+    modToNames: Map[String, Map[String, String]],
+    renameMap: RenameMap,
+    ct: CircuitTarget, igraph: InstanceKeyGraph
+  )(mod: DefModule): DefModule = {
+
+    val mt = ct.module(mod.name)
+    val instToOfModule = mutable.HashMap[String, String]()
+    val names = modToNames.getOrElse(mod.name, Map.empty[String, String])
+    // Need to find WRef referring to mems for prefixing
+    def onExpr(expr: Expression): Expression = expr match {
+      case wref @ WRef(n, _,_,_) if names.contains(n) =>
+        wref.copy(name = names(n))
+      case sub @ SubField(WRef(i, _, _, _), p,_,_) if instToOfModule.contains(i) =>
+        val newsub = modToNames.get(instToOfModule(i)) match {
+          case Some(map) if map.contains(p) => sub.copy(name = map(p))
+          case _ => sub
+        }
+        newsub.map(onExpr)
+      case other => other.map(onExpr)
+    }
+    def onStmt(stmt: Statement): Statement = stmt match {
+      // Yes we match on instance name to rename the module
+      case inst: DefInstance if names.contains(inst.name) =>
+        instToOfModule(inst.name) = inst.module
+        val newName = names(inst.name)
+        renameMap.record(mt.instOf(inst.name, inst.module), mt.instOf(newName, inst.module))
+        inst.copy(name = names(inst.name))
+      case inst: DefInstance =>
+        instToOfModule(inst.name) = inst.module
+        inst
+      case error @ (_: WDefInstanceConnector) =>
+        throw new Exception(s"Unexpected instance object $error")
+      // Prefix SeqMems because they result in Verilog modules
```

Comment:
As `WDefInstanceConnector` is kind of a FIRRTL implementation detail, is this necessary?